File: ToolClass/OC/Category/UIButton+Extension.m
//
//  UIButton+Extension.m
//  SH
//
//  Created by i7colors on 2019/9/4.
//  Copyright © 2019 surhoo. All rights reserved.
//

#import "UIButton+Extension.h"
#import <objc/runtime.h>

typedef void(^ButtonEventsBlock)(void);
static const NSString *Key_TouchEdgeInsets = @"touchEdgeInsets";
@interface UIButton()
/** 事件回调的block */
@property (nonatomic, copy) ButtonEventsBlock buttonEventsBlock;
@end

@implementation UIButton (Extension)
- (void)layoutWithEdgeInsetsStyle:(ButtonEdgeInsetsStyle)style
                        imageTitleSpace:(CGFloat)space {
    /**
     *  知识点：titleEdgeInsets是title相对于其上下左右的inset，跟tableView的contentInset是类似的，
     *  如果只有title，那它上下左右都是相对于button的，image也是一样；
     *  如果同时有image和label，那这时候image的上左下是相对于button，右边是相对于label的；title的上右下是相对于button，左边是相对于image的。
     */
    
    // 1. 得到imageView和titleLabel的宽、高
    CGFloat imageWith = self.imageView.image.size.width;
    CGFloat imageHeight = self.imageView.image.size.height;
    
    CGFloat labelWidth = 0.0;
    CGFloat labelHeight = 0.0;
    if ([UIDevice currentDevice].systemVersion.floatValue >= 8.0) {
        // 由于iOS8中titleLabel的size为0，用下面的这种设置
        labelWidth = self.titleLabel.intrinsicContentSize.width;
        labelHeight = self.titleLabel.intrinsicContentSize.height;
    } else {
        labelWidth = self.titleLabel.frame.size.width;
        labelHeight = self.titleLabel.frame.size.height;
    }
    if (!CGSizeEqualToSize(self.titleLabel.frame.size, CGSizeZero)) {
        labelWidth = self.titleLabel.frame.size.width;
        labelHeight = self.titleLabel.frame.size.height;
    }
    
    // 2. 声明全局的imageEdgeInsets和labelEdgeInsets
    UIEdgeInsets imageEdgeInsets = UIEdgeInsetsZero;
    UIEdgeInsets labelEdgeInsets = UIEdgeInsetsZero;
    
    // 3. 根据style和space得到imageEdgeInsets和labelEdgeInsets的值
    
    switch (style) {
        case ButtonEdgeInsetsStyleTop:
        {
            imageEdgeInsets = UIEdgeInsetsMake(-labelHeight-space/2.0, 0, 0, -labelWidth);
            labelEdgeInsets = UIEdgeInsetsMake(0, -imageWith, -imageHeight-space/2.0, 0);
        }
            break;
        case ButtonEdgeInsetsStyleLeft:
        {
            imageEdgeInsets = UIEdgeInsetsMake(0, -space/2.0, 0, space/2.0);
            labelEdgeInsets = UIEdgeInsetsMake(0, space/2.0, 0, -space/2.0);
        }
            break;
        case ButtonEdgeInsetsStyleBottom:
        {
            imageEdgeInsets = UIEdgeInsetsMake(0, 0, -labelHeight-space/2.0, -labelWidth);
            labelEdgeInsets = UIEdgeInsetsMake(-imageHeight-space/2.0, -imageWith, 0, 0);
        }
            break;
        case ButtonEdgeInsetsStyleRight:
        {
            imageEdgeInsets = UIEdgeInsetsMake(0, labelWidth+space/2.0, 0, -labelWidth-space/2.0);
            labelEdgeInsets = UIEdgeInsetsMake(0, -imageWith-space/2.0, 0, imageWith+space/2.0);
        }
            break;
        default:
            break;
    }
    
    // 4. 赋值
    self.titleEdgeInsets = labelEdgeInsets;
    self.imageEdgeInsets = imageEdgeInsets;
}

#pragma mark - button的block
//------- 添加属性 -------//

static void *my_buttonEventsBlockKey = &my_buttonEventsBlockKey;

- (ButtonEventsBlock)buttonEventsBlock {
    return objc_getAssociatedObject(self, &my_buttonEventsBlockKey);
}

- (void)setButtonEventsBlock:(ButtonEventsBlock)buttonEventsBlock {
    objc_setAssociatedObject(self, &my_buttonEventsBlockKey, buttonEventsBlock, OBJC_ASSOCIATION_COPY);
}

- (void)addEventHandler:(void (^)(void))block {
    self.buttonEventsBlock = block;
    [self addTarget:self action:@selector(addEventHandler) forControlEvents:UIControlEventTouchUpInside];
}

// 按钮点击
- (void)addEventHandler {
    !self.buttonEventsBlock ?: self.buttonEventsBlock();
}

- (void)setBtnWithTitle:(NSString *)title titleColor:(UIColor *)color font:(CGFloat)font {
    [self setTitle:title forState:UIControlStateNormal];
    [self setTitleColor:color forState:UIControlStateNormal];
    self.titleLabel.font = [UIFont systemFontOfSize:font];
}

#pragma mark - 扩大点击区域
- (BOOL)pointInside:(CGPoint)point withEvent:(UIEvent *)event
{
    if(UIEdgeInsetsEqualToEdgeInsets(self.touchEdgeInsets, UIEdgeInsetsZero) || !self.enabled || self.hidden)
    {
        return [super pointInside:point withEvent:event];
    }
    
    CGRect relativeFrame = self.bounds;
    CGRect hitFrame = UIEdgeInsetsInsetRect(relativeFrame, self.touchEdgeInsets);
    
    return CGRectContainsPoint(hitFrame, point);
}

- (void)setTouchEdgeInsets:(UIEdgeInsets)touchEdgeInsets {
    UIEdgeInsets edges = UIEdgeInsetsMake(-touchEdgeInsets.top, -touchEdgeInsets.left, -touchEdgeInsets.bottom, -touchEdgeInsets.right);
    NSValue *value = [NSValue value:&edges withObjCType:@encode(UIEdgeInsets)];
    objc_setAssociatedObject(self, &Key_TouchEdgeInsets, value, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (UIEdgeInsets)touchEdgeInsets {
    NSValue *value = objc_getAssociatedObject(self, &Key_TouchEdgeInsets);
    if(value){
        UIEdgeInsets edgeInsets;
        [value getValue:&edgeInsets];
        return edgeInsets;
    } else {
        return UIEdgeInsetsZero;
    }
}

@end

File: ToolClass/Swift/Extension/UIButton+Extension.swift
//
//  UIButton+Extension.swift
//  ToolClass
//
//  Created by caidongdong on 2021/6/24.
//

import Foundation

public extension UIButton {
    
    @objc enum ButtonImageEdgeInsetsStyle: Int {
        case top, left, right, bottom
    }
    
    @objc func imagePosition(at style: ButtonImageEdgeInsetsStyle, space: CGFloat) {
        imagePosition(at: style, space: space, isAutoWidth: false , maxWidth: -1.0)
    }
    
    @objc func imagePosition(at style: ButtonImageEdgeInsetsStyle, space: CGFloat, isAutoWidth: Bool = false) {
        imagePosition(at: style, space: space, isAutoWidth: isAutoWidth , maxWidth: -1.0)
    }
    
    @objc func imagePosition(at style: ButtonImageEdgeInsetsStyle, space: CGFloat, maxWidth: CGFloat = -1.0) {
        imagePosition(at: style, space: space, isAutoWidth: false , maxWidth: maxWidth)
    }
    
    @objc func imagePosition(at style: ButtonImageEdgeInsetsStyle, space: CGFloat, isAutoWidth: Bool = false, maxWidth: CGFloat = -1.0) {
        guard let imageV = imageView else { return }
        guard let titleL = titleLabel else { return }
        self.titleLabel?.lineBreakMode = .byTruncatingTail
        //获取图像的宽和高
        let imageWidth = imageV.frame.size.width
        let imageHeight = imageV.frame.size.height
        //获取文字的宽和高
        var labelWidth  = titleL.intrinsicContentSize.width
        var labelHeight = titleL.intrinsicContentSize.height
        var imageEdgeInsets = UIEdgeInsets.zero
        var labelEdgeInsets = UIEdgeInsets.zero
        
        if #available(iOS 8, *) {
            labelWidth = titleL.intrinsicContentSize.width
            labelHeight = titleL.intrinsicContentSize.height
        } else {
            labelWidth = titleL.frame.size.width
            labelHeight = titleL.frame.size.height
        }
        
        if maxWidth > 0 {
            let max_labWidth = maxWidth - imageWidth - space
            if labelWidth > max_labWidth {
                labelWidth = titleL.frame.size.width
            }
        }
        
        switch style {
        case .left:
            //正常状态--只不过加了个间距
            imageEdgeInsets = UIEdgeInsets(top: 0, left: -space * 0.5, bottom: 0, right: space * 0.5)
            labelEdgeInsets = UIEdgeInsets(top: 0, left: space * 0.5, bottom: 0, right: -space * 0.5)
            if isAutoWidth == true {
                snp.makeConstraints { m in
                    m.width.equalTo(intrinsicContentSize.width + space)
                }
            }
        case .right:
            //切换位置--左文字右图像
            imageEdgeInsets = UIEdgeInsets(top: 0, left: labelWidth + space * 0.5, bottom: 0, right: -labelWidth - space * 0.5)
            labelEdgeInsets = UIEdgeInsets(top: 0, left: -imageWidth - space * 0.5, bottom: 0, right: imageWidth + space * 0.5)
            if isAutoWidth == true {
                snp.makeConstraints { m in
                    m.width.equalTo(intrinsicContentSize.width + space)
                }
            }
        case .top:
            //切换位置--上图像下文字
            imageEdgeInsets = UIEdgeInsets(top: -labelHeight - space * 0.5, left: 0, bottom: 0, right: -labelWidth)
            labelEdgeInsets = UIEdgeInsets(top: 0, left: -imageWidth, bottom: -imageHeight - space * 0.5, right: 0)
        case .bottom:
            //切换位置--下图像上文字
            imageEdgeInsets = UIEdgeInsets(top: 0, left: 0, bottom: -labelHeight - space * 0.5, right: -labelWidth)
            labelEdgeInsets = UIEdgeInsets(top: -imageHeight - space * 0.5, left: -imageWidth, bottom: 0, right: 0)
        }
        
        self.titleEdgeInsets = labelEdgeInsets
        self.imageEdgeInsets = imageEdgeInsets
    }
    
    
}

@objc
extension UIButton {
    private struct AssociatedKeys {
        static var touchInsetsKey = UIEdgeInsets.zero
    }
    
    var touchInsets: UIEdgeInsets {
        get {
            return objc_getAssociatedObject(self, &AssociatedKeys.touchInsetsKey) as? UIEdgeInsets ?? .zero
        }
        set {
            objc_setAssociatedObject(self, &AssociatedKeys.touchInsetsKey, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
    ///如果失效的话,监察是否有其他地方重写了这个方法
    open override func point(inside point: CGPoint, with event: UIEvent?) -> Bool {
        let largerRect = bounds.inset(by: touchInsets.inverted())
        return largerRect.contains(point)
    }
}

extension UIEdgeInsets {
    func inverted() -> UIEdgeInsets {
        return UIEdgeInsets(top: -top, left: -left, bottom: -bottom, right: -right)
    }
}
